designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,275 @@
+```
+
+### Each suggestion object requires the `output` property
+
+In addition to the message provided by the suggestion, the actual code output is another key part of the suggestion that tests should require:
+
+```pt
+AssertionError [ERR_ASSERTION]: Test suggestion object is missing 'output`.
+```
+
+### Disallow repeating the test case `code` as `output` for non-fixable test cases
+
+Today, a non-fixable invalid test case can be written like this:
+
+```js
+{
+    code: 'console.log("foo");',
+    output: 'console.log("foo");',
+    // ...
+}
+```
+
+This is not ideal, especially when the code is longer and multi-line:
+
+- It's not immediately clear whether the `code` differs from the `output`, making it difficult to determine whether or not the test case produces an autofix
+- The maintenance burden is higher as changes to the test case need to be replicated in both `code` and `output`
+- The test case is less concise, needlessly increasing the size/length of the test file
+
+So we will disallow repeating the test case `code` as `output` for non-fixable test cases using a new assertion:
+
+```pt
+AssertionError [ERR_ASSERTION]: Test error object 'output' matches 'code'. Omit 'output' for non-fixable test cases.
+```
+
+To clearly and concisely indicate that a test case produces no autofix, it is recommended to omit the `output` property entirely. Note that we will also allow `output: null` or `output: undefined` which can be useful when dynamically generating test cases (e.g. `output: hasAutofix ? autofixedCode : null`), and because many existing test cases are written with `output: null` as that used to be necessary to assert that the test case had no autofix (it isn't necessary anymore as of [ESLint v7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class)).
+
+An existing lint rule [eslint-plugin/prefer-output-null](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-output-null.md) enforces the desired behavior and is autofixable which should help with migration.
+
+### Disallow identical test cases
+
+Duplicate test cases can cause confusion, can be hard to detect manually in a long file, and serve no purpose.
+
+We will add an assertion when a duplicate test case is detected:
+
+```pt
+AssertionError [ERR_ASSERTION]: Detected duplicate test case.
+```
+
+As test cases end up represented as objects, the problem here is essentially to detect duplicate objects in an array. To handle this, here's one efficient algorithm (also mentioned in this [article](https://medium.com/programming-essentials/how-to-know-if-an-array-has-duplicates-in-javascript-27d99ab9a0d2)):
+
+1. Maintain a set of the test cases we have seen so far.
+2. For each test case, use `JSON.stringify()` to get a string representation of the test case object. Note that the object keys should be sorted before serializing.
```

Comment:
Instead of sorting keys manually (which should be ideally done deeply, for every object encountered during the serialization), we could use a library like `json-stable-stringify-without-jsonify`.  We already have it as a dependency.

Reply:
Good point, updated to use `json-stable-stringify-without-jsonify` instead.